### ISSUE TITLE
TEE Enabled Bacalhau prototype, testing, and research

### ISSUE THREAD

**Issue by PHABRIC77** [filecoin-project/devgrants#1146]
# Open Grant Proposal: `TEE Enabled Bacalhau prototype, FVM proof, FVM linked KMS (optional)

**encloud technology:**

**Proposal Category:** `core-dev`, 

**Proposer:** `Phabric77, parthbond180`

**(Optional) Technical Sponsor:** `David Aronchick, Sonia John. Support from Wes Floyd, Luke Marsden, Simon Worthington`

**Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT, APACHE2, or GPL licenses?:** Yes

# Project Description

Data privacy in use is critical to the effective deployment of sensitive data. Enterprise data is often inherently sensitive and valuable, but at present there is no way to perform compute whilst preserving privacy. This is a major barrier to the adoption at scale of Filecoin/IPFS as a data storage solution for enterprise data. Enterprise data is an area of key focus for the Filecoin network and this focus is typified by the imminent introduction of the Enterprise FIL+ incentive program, building on the success of the FIL+ program for public data.

Data privacy in use is solved by a number of emerging technologies like homomorphic encryption and confidential computing. A key priority for Filecoin & Protocol Labs is building a compute layer for IPFS, and Bacalhau are the Protocol Labs sponsored project responsible for building a compute layer. encloud (formerly 180Protocol/Bond180) have identified a need for privacy enhanced compute for Bacalhau. In encloud’s view, privacy and private compute can reduce the trust barrier preventing mass scale adoption of decentralized cloud.

As part of Open Grant #631 encloud undertook a research project to assess the options for delivering privacy enhanced compute. Based on a detailed assessment of the options encloud presented a recommendation at the COD Summit in Lisbon. Following the delivery of this recommendation encloud has been requested by the interested and sponsoring stakeholders to submit a grant to build and test a privacy enhanced compute enhancement for Bacalhau. The recommendation was to use trusted execution environments (TEE’s) to provide privacy enhanced compute and accordingly this grant proposes a prototype to deliver privacy enhanced compute for the Bacalhau network.

Currently, encloud leverages encryption for data at rest and confidential computing to aggregate and compute on sensitive data sets. encloud has already enabled storage of encrypted data inputs and outputs from confidential computation on sensitive and commercially valuable data on IPFS, and will enhance this capability by building single party privacy enhanced compute for Filecoin/IPFS by extending the open source Bacalhau compute platform. Next steps would be to build multi-party compute.

An important feature to attract new data sets onto Filecoin is to enable computation on encrypted data. Confidential computing refers to emerging technologies that enable privacy assured and secure use of sensitive encrypted data sets.
Once stored on the network, sensitive data can be leveraged by sharing, aggregating, and computing. By providing tools to securely leverage sensitive data, value derived from the data is enhanced thereby incentivizing data owners to deploy such data.

“Confidential Computing protects data in use by performing computation in a hardware-based Trusted Execution Environment ([Confidential Computing Consortium](https://confidentialcomputing.io/), 2022)”.

Trusted Execution Environments (TEE’s) offer the most scalable, usable and performant means to perform privacy enhanced computations. This may change in the coming years with the emergence of scalable fully homomorphic encryption. 
Additionally, privacy enhancing technologies combined with blockchain ensures both traceable data contributions alongside privacy assurance, meaning no single actor can gain preferential access to data that is used. The combination of these technologies is a powerful enabler for single and multi party computations on sensitive data.

These capabilities when combined with FVM, for example, are transformational for the ecosystem and will enable a plethora of use cases leveraging sensitive data in an algorithmic way.

## Value

Our journey began as a Fintech, and over 2 years we learned that banks and investment managers were unwilling to share sensitive data. We built encloud to solve this problem, through privacy technology and decentralization we aimed to eliminate the risk of sharing and computing on sensitive data.
 
Our vision is a privacy enabled collaborative ecosystem with little or no reliance on central actors, enabled by FVM. In this future state, data owners will be able to commercialize their data fairly and securely. We believe this will enable a vast swathe of under-commercialized sensitive data to be leveraged effectively and prove a marquee use case for the decentralized cloud. encloud thus unlocks value in sensitive data. Sensitive data assets include, for example, databases, documents, pictures, spreadsheets, etc., that carry data and which have strategic or operative value. These data assets have value and could generate commercial returns for owners, but they are not effectively deployed.

Furthermore, Web2 Cloud Service Provider (CSPs) in the past two years have started offering TEE based confidential VMs as a means to attract workloads from privacy sensitive customers. Web2 CSPs often claim that these TEE based confidential VMs offer complete isolation and risk mitigation for sensitive data - including from the CSP itself. However, this is questionable as in a lot of these early offerings the CSP modifies core TEE based workflows (attestation, deployment) to fit CSP specific designs that often require the CSP to be trusted along with the TEE chip manufacturer. 

Ideally with usage of TEEs the only root of trust should be the chip manufacturer, i.e., Intel or AMD, rather than a specific CSP. Creating a non-CSP based TEE infrastructure enables privacy sensitive customers to avoid having to trust the CSP with TEE based workloads. A TEE based Bacalhau would ideally run on the various independent Storage Providers (SPs) on the Filecoin network. These SPs would offer privacy enhanced compute and the customer would be able to verify that the SP is running a genuine TEE before deploying their encrypted sensitive workloads onto the Bacalhau computation layer.

Deploying sensitive workloads to the Filecoin network can unlock tremendous value. Consider that  the majority of the value of the S&P 500 is in the form of data assets. encloud believes the Filecoin network can compete with existing Web2 CSPs to take part in the coming revolution of enterprise.

Today the Bacalhau network has processed roughly 28k jobs, a majority of these with Dockerized based workloads. From our interviews with the core Bacalhau dev team and Wes, the commercial lead, it has become clear that compute on encrypted data is a key feature desired by the ecosystem and privacy focused customers. Today bacalhau is hosted on GCP cloud and will be for the near to medium term. 

GCP supports confidential VMs with AMD EPYC CPU’s and SEV capability for compute on encrypted data. However configuring and managing these VMs is quite cumbersome for users without understanding the intimacies of confidential computing. Encloud will create a tool that makes deployment of dockerized workloads inside GCP confidential VM a one command task abstracted under the Bacalhau CLI, allowing clients to run compute on encrypted data stored on IPFS. 

By building FVM integration for private compute jobs and optionally for key management encloud will integrate Bacalhau with FVM opening a litany of FVM derived privacy orientated use cases such as secure data daos, marketplaces and execution venues for digital assets and/or jobs, and privacy assured staking protocols.

Today key custody is a major challenge in the ecosystem. For privacy enhanced compute, encryption keys have to be granted to the secure enclave compute unit based on verification of the enclaves identity and workload deployed on it. Such a key custody and policy manager can be developed on web2 Key Vault tooling like HashiCorp vault. However, this requires key owners to host web2 infra and rely on centralized KMS tooling. With the FEVM, smart contracts can be used to custody keys and release access based on validation logic. This reduces the friction of operating web2 infrastructure for key custody.

This can drastically enhance adoption of Bacalhau as a compute tool for privacy sensitive customers. 

From our understanding there is a desire for all bacalhau workloads to be run in this manner and can fundamentally shape the way privacy is considered by the IPFS ecosystem.  

## Risks 

A risk is the emergence of cryptographic techniques or quantum computational technology that can override and decrypt modern cryptographic protocols utilized to encrypt sensitive data. To this end, quantum resistant cryptographic protocols can be utilized to encrypt sensitive data prior to storing on Filecoin/IPFS. 

Another challenge remains the reliance on external libraries and vendors. Bacalhau and the Filecoin VM are under development. Dependency on third parties increases operating risks. To mitigate these risks, encloud will  will develop its own technology to diminish this risk.

Ensuring the data stored on IPFS is useful, deployable, and utilized is likely to be a key ambition for Filecoin and our intention is to align where possible with other services and incentive based systems to maximize impact for the network and mitigate these risks.

## Deliverables

encloud’s solution under this grant proposal covers: -
 
**Technical Deliverables**
 
- Encloud have already built an open-source encryption and storage CLI and an desktop/web application that both leverage envelope encryption (the standard for encryption employed Amazon, Google, Microsoft etc), non-custodial key management, file sharing and retrieval 

- encloud also want to enable open-source private computations over data. 

- encloud further want to provide a mechanism for providers to verify private compute jobs using FVM. Once a job is executed, a proof for execution could be posted on chain to claim rewards/enhance provider reputation.

- Deliverables are a design and prototype of an open-sourced TEE based Bacalhau for single party compute, and a design and prototype for FVM integration for the proof mechanism described above

- Next steps include expanding to enable multi-party compute targeting privacy for federated use cases.

- Focus on single party privacy enhanced compute integration 
   - Use open-source TEE deployment and verification technology to deploy client workloads as a K8 cluster inside TEEs. Currently the open source TEE technology requires the use of GCP to deploy TEE VMs. The aim will be to move to a bare metal deployment in the future (not in this grant scope)
  
 
- Sensitive data will be encrypted and stored on IPFS/Filecoin by the user/data owner. The user will be responsible for the custody and management of the encryption keys. Sensitive data will be picked up by the Executor nodes and decrypted by the TEE during job execution. 

## Development Roadmap

encloud want to enable secure computations on sensitive data sets. Based on our work from Grant 631, we have tested out various privacy enhanced computation approaches and assessed their fit for the Bacalhau network. As a result of design, development, testing and benchmarking of Fully Homomorphic Encryption and Trusted Execution Environments based approaches we recommend following a TEE based Bacalhau.

The objective of this grant will be to develop the architectures and prototypes from Grant 631 and extend the Bacalhau tooling to add support for computation on sensitive data using TEEs. Additionally we will integrate with FVM for a proofing and optional KMS solution. To this end the following will be done -

**Milestone 1: Milestone 1: Design an open-sourced GCP based Bacalhau for single party compute**

[$2,5k]

- Design of a GCP based Bacalhau prototype using a simple workload deployed on an AMD SEV TEE using the Bacalhau CLI

- Preparing the GCP confidential VM image and hosting on a GCP image registry. Deploy the confidential VM image on a GCP AMD EPYC Confidential VM with SEV support

- Integrating the GCP VM deployment and image registry commands with Bacalhau CLI as a new command and relevant server side changes to Requestor, Compute, & Executor interfaces

 **Milestone 2: Development of an open-source prototype Bacalhau for private single party compute**

[ $15k]

- Post development of TEE based Bacalhau prototype development work will commence with close coordination with the Bacalhau core development team to deliver a fully integrated and testnet ready product capable of deploying any dockerized workload supported by a non TEE Bacalhau

- The aim will be to make regular check ins to the Bacalhau open source repository 

- Unit testing, appropriate documentation and “how to use” guides will be delivered as part of the development tasks in line with best development best practices

- Focus on a simple **user experience** will be key. The aim will be to keep the modified Bacalhau specific CLIs as simple for the user as possible

**Milestone 3 (optional for encloud): Design an open-sourced proof of execution leveraging FVM integration**

[$5k]

A key challenge with decentralized compute is verifiability and with enclaves native attestation capabilities and the FVM smart contracts compute providers can write proofs of executions on chain for verifiers to validate.

- Design a mechanism to allow proofs of executions to be written onto the Filecoin chain

- FVM actors for compute providers that write proof of execution from within the secure enclave to the chain

- FVM actor for verifier that validate these proof of executions and allocate rewards/reputation to the provider for successful execution

**Milestone 4 (optional for encloud): Development of a fully integrated open-source open-sourced proof of execution leveraging FVM integration**

[$12.5k]

- Develop smart contracts for a privacy enhanced compute provider FVM actor that posts proof of executions for bacalhau compute job executed inside a secure enclave

- Proofs can include the enclave attestation report validating enclave genuineness

- Proofs also includes a workload hash confirming that the right workload was executed inside the enclave

- Develop smart contracts for a compute verifier FVM actor that can validate proof of executions posted by compute providers for privacy enhanced bacalhau compute jobs by validating the enclave attestation and the workload hash 

- Deploy actor smart contracts to FVM Hyperspace testenet and test with bacalhau jobs

**Milestone 5 (optional):  Optional deliverable for both FF and encloud- Design and build encryption Key management using FVM**

[$10k]

- Utilize FEVM smart contracts to custody encryption keys by leveraging existing tooling like lit protocol that support key custody using EVM chain based smart contracts 

- Develop smart contracts that provision encryption keys to a secure enclave based on attestation requests from the enclave 

## Total Budget Requested

**$35k** (with optional 10k increase to $45k)

## Maintenance and Upgrade Plans

As a foundational project, the durability of the platform is critical to the success of what is a long-term project. Envisaging a wide base of users, maintenance of the platform is very important to us.

Once the outlined milestones have been completed, a detailed plan for long-term maintenance will be drafted by the team and stakeholders. Updates to the platform as per our project roadmap, will be made to stay relevant to our ever-changing needs. Wider development of traction and use cases, and further integration with decentralized technologies are the main pillars of platform upgrades.

# Team

## Team Members

- Phil Holbrook, CEO: https://www.linkedin.com/in/philip-holbrook/
- Parth Shukla, CTO: https://github.com/parthbond180, https://www.linkedin.com/in/parthshukla1/
- Vivek Shah, Senior Developer, https://github.com/vivekshah92
- Hitesh Patel, Senior Developer
https://github.com/hiteshvpatel256
- Deepanshu Arora, Junior developer, https://github.com/Deepansharora27
- Igor Lobanov, Technology Architecture Advisor, https://github.com/lobanov, https://github.com/igor-b180

**Advisory board**
- Keith Bear, Enterprise Blockchain Advisor - https://www.linkedin.com/in/keith-bear-2b7407
- Ian Hunt, Financial Technology advisor - https://www.linkedin.com/in/ian-hunt-228880

## Team Member LinkedIn Profiles

Website: https://encloud.tech/

## Relevant Experience

The core team spun out of the University of Cambridge to found the FinTech startup Bond180. Bond180, a technology company providing technology solutions for fixed income markets, received early financial backing from R3 and the University of Cambridge and were hosted at the Outlier Ventures Basecamp.

encloud has spawned from that group, widening the technology’s scope and redirecting its focus to the Web3.0 landscape. Our team benefits from its eclectic background in our capacity to innovate. 

The encloud team is composed of passionate individuals and creative problem solvers aiming to make enterprise on ramp to web3. We are on a mission to unlock the value in sensitive enterprise data through the deployment of confidential computing and DLT. The core team is composed of Phil Holbrook, and Parth Shukla, who established Bond180, the tech service provider for encloud. Phil has 15 years experience working on the buy-side of finance in the City of London and Parth is an experienced enterprise architect and developer, with expertise in R3 Corda, and Ethereum in terms of Layer 1 and experience with several TEE frameworks like Conclave, Apache Teaclave, Anjuna etc. The team is supported by two senior developers based out of India, both with experience in writing enterprise grade distributed applications. The team is advised by Igor Lobanov, former Chief Architect at Legal and General Investment Management, Keith Bear, formerly Head of Blockchain at IBM and Ian Hunt, a buy-side technology expert. Further, encloud has Leo Greve and Chiel Ruiter as NED’s, both experienced capital markets bankers with several decades of experience across European financial markets.

## Team code repositories

GitHub: https://github.com/encloud & https://github.com/180Protocol

# Additional Information

We were previously recipients of a open grant [Open Grant Proposal: encloud · Issue #451 · filecoin-project/devgrants · GitHub]. & [Open Grant Proposal: encloud · Issue #631 · filecoin-project/devgrants · GitHub].

Please contact [phil@](mailto:phil@bond180.com)encloud.tech  & [parth@](mailto:parth@bond180.com)encloud.tech to discuss
#1099

**Comment by lukemarsden** [comment]
LGTM 👍

**Comment by wesfloyd** [comment]
For the Bacalhau team's perspective - this grant would open up additional use cases for Bacalhau processing of sensitive data stored in Filecoin. For example, Healthcare and Financial Services datasets often require compute over encrypted data. This grant work would be the first capability for Bacalhau to service those workloads.

**Comment by lukemarsden** [comment]
Hi @ErinOCon, I'm the eng lead on the Bacalhau project. The TEE team have completed their first grant. Please reconsider. We are very keen to see development on this grant get underway, as it unblocks important new use cases for us! Thanks!

cc @aronchick

**Comment by PHABRIC77** [comment]
> Hi @PHABRIC77, thank you for your proposal! We would like the grant that is currently active to resolve before considering this project further. Once the final milestone is complete, please feel welcome to resubmit your proposal.
> 
> If/when resubmitted, we would be interested to learn more about your sustainability plans (building user traction, community engagement, etc.)

HI, we have submitted the grant deliverables on a call with @wesfloyd on @aronchick 's and @realChainLife . We shared the recording of the check in call and also the report that we authored with the results of the milestone. Could you please re-open this grant, and we will add any supporting and additional information as you need regarding sustainability/user traction etc.

**Comment by ErinOCon** [comment]
Hi @PHABRIC77, let's make sure we are all on the same page. I've started a Slack thread to discuss the progress of Issue #451.

@lukemarsden, thank you for your reply! I will be in touch shortly with an update.

**Comment by parthbond180** [comment]
> Hi @PHABRIC77, thank you for your proposal! We would like the grant that is currently active to resolve before considering this project further. Once the final milestone is complete, please feel welcome to resubmit your proposal.
> 
> If/when resubmitted, we would be interested to learn more about your sustainability plans (building user traction, community engagement, etc.)

Hi Erin, 

Thanks for your question regarding user traction and community engagement. We recognize that evidence of user traction is key towards committing investment towards this grant. We are pleased to share that over the past two months we have committed extensive due dilligence with various participants in the ecosystem who have all echoed the need for privacy technologies as a means of onboarding more users to the Filecoin ecosystem. 

We participated in ESPA as part of the 4th cohort where we got to understand the intricacies of the economics that motivate Storage Providers in the ecosystem. Here we learned that SPs increasingly need additional revenue beyond basic storage, by providing add on services like compute. Compute is being built by the Bacalhau project. Additionally, SPs are targeting enterprises to unlock a new wave of user onboarding into the ecosystem. However, enterprises require privacy as a default for any interaction with Filecoin. 

Hence, enabling privacy enhanced compute is important for SPs to onboard enterprise users to run compute on the network. We corroborated this with various SPs in the ecosystem and have POCs lined with a few as well to test out this compute layer.

Additionally, we presented the idea for this privacy enhanced compute layer during the FIL Lisbon summit in November, where we received additional validation on use cases from the ecosystem. Post my talk during the COD days of the summit, I was approached by various DeSci startups who are interested in running compute on sensitive data sets. We have also interest expressed from other web3 compute layers and protocols who see the benefit of privacy compute for use cases like decentralized MFA.

Beyond validation from within the ecosystem, we have also reached out to various enterprises architects who all agree that a privacy compute layer will provide key missing elements in the Filecoin and broadly the Web3 ecosystems.

The architecture of the proposed compute layer has been corroborated with the core Bacalhau dev team and there is a strong interest expressed from both Bacalhau and CoD for this feature. Post delivery of M2 of this grant we want to conduct joint marketing and go to market to onboard compute workloads to the testnet. Our team is already following a similar approach for our Encloud encryption tooling (built on Estuary) in conjunction with the Estuary team.

With such strong validation and expressed interest from stakeholders of every kind, we hope we can build this feature out as we have all parts of the puzzle in place, including the architecture, vendor partnerships, POCs.
